### PR TITLE
Implement weighted sampling

### PR DESCRIPTION
**Pull request by victorlin**
## Description of proposed changes

This new option takes a weights file to determine group sizes. Summarized by the help text:

https://github.com/nextstrain/augur/blob/9a2090b1de3227e8267189b5f93ad6880f29f74a/augur/filter/__init__.py#L73-L97

## Related issue(s)

- #1318

## Notable comment threads

- [x] [How to handle a mix of weighted and unweighted columns?](https://github.com/nextstrain/augur/pull/1454#discussion_r1596214631) Apply weights within unweighted groups
- [x] [How to handle weighted groups that don't appear in metadata?](https://github.com/nextstrain/augur/pull/1454#discussion_r1596227089) Drop them
- [x] [Add an option to show number of outputted sequences?](https://github.com/nextstrain/augur/pull/1454#discussion_r1607249501) Done
- [x] [How to handle over-targeting group sizes with only weighted columns?](https://github.com/nextstrain/augur/pull/1454#discussion_r1704761006) Added warnings in output
- [x] [How to handle over-targeting group sizes with a mix of weighted and unweighted columns?](https://github.com/nextstrain/augur/pull/1454#discussion_r1704757415) Added warnings in output
- [x] [Clarify interaction with probabilistic sampling](https://github.com/nextstrain/augur/pull/1454#discussion_r1723933707) Done

## Checklist

- [x] Prototype usage in an existing workflow: https://github.com/nextstrain/ncov/pull/1106
- [x] Automated checks pass
- [x] [Check][1] if you need to add a changelog message
- [x] [Check][2] if you need to add tests (added)
- [x] [Check][3] if you need to update docs
- [x] Create issue for [FIXME](https://github.com/nextstrain/augur/blob/918032f659f3107913efe285dc6ee2e5afb9fcaa/augur/filter/validate_arguments.py#L58): https://github.com/nextstrain/augur/issues/1590

[1]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#updating-the-changelog
[2]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#testing
[3]: https://github.com/nextstrain/augur/blob/-/docs/contribute/DEV_DOCS.md#when-to-update

### CODE REVIEW

File: augur/filter/subsample.py [hunk 1]
```diff
@@ -298,6 +299,25 @@ def get_group_sizes(groups, target_group_size, random_seed=None):
     return max_sizes_per_group
 
 
+def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size):
+    weights = pd.read_csv(weights_file, delimiter='\t')
+
+    # FIXME: check that weights file covers all `groups`
```
Comment (jameshadfield) [inline augur/filter/subsample.py]: _(Notes while skimming the draft PR)_

I'd expect a common usage would be a weights TSV covering a subset of the groups. E.g. provide weights for country and then `--group-by country year` or somesuch. Do you think this is reasonable to support in the v1?

Comment (victorlin) [reply augur/filter/subsample.py]: Yes, what you've described is in the current implementation with a small difference in the command line configuration.

https://github.com/nextstrain/augur/blob/adc1d0bde9ab1832197f64060a924f1f43301971/tests/functional/filter/cram/subsample-weighted.t#L48-L67

I'll look into updating to require all columns used for grouping in `--group-by` rather than just the unweighted columns - that seems more clear.

Comment (victorlin) [reply augur/filter/subsample.py]: Done as 8d72be055d1edcaf2c4a411c4757f9dc9dd655a3

**Comment by victorlin** [comment]
This should be in a working state, at least based on limited testing in [subsample-weighted.t](https://github.com/nextstrain/augur/blob/c7da4c72a1b925615c492e8fb4eec99ce8a05c30/tests/functional/filter/cram/subsample-weighted.t). Planning to test using #1444 + https://github.com/nextstrain/ncov/pull/1102 to get a feel for how add an additional weights.tsv file fits in a pathogen repo and make sure it works as expected on real data.

File: augur/filter/subsample.py [hunk 2]
```diff
@@ -297,6 +299,69 @@ def get_probabilistic_group_sizes(groups, target_group_size, random_seed=None):
...
+    if missing_groups:
+        error = "The following groups appear in the metadata but are missing from the weights file:"
+        for group in sorted(missing_groups):
+            column_value_pairs = [f"{column}={value!r}" for column, value in zip(group_by, group)]
+            error += f"\n\t{', '.join(column_value_pairs)}"
+        raise AugurError(error)
+
+    # Calculate maximum group sizes based on weights
+    SIZE_COLUMN = '_augur_filter_target_size'
+    weights[SIZE_COLUMN] = weights[WEIGHTS_COLUMN] / weights[WEIGHTS_COLUMN].sum() * target_total_size
```
Comment (victorlin) [inline augur/filter/subsample.py:360]: Testing in https://github.com/nextstrain/ncov/pull/1106 revealed that this isn't the right calculation. It generates fractions but integers are needed. I noticed because the gisaid/asia/all-time run produced 4391 sequences for the focal sample despite specifying a max of 3500.

Something similar to what's done with `--probabilistic-sampling` might be necessary here, which takes a fractional sequences per group and determines integer sizes.

Comment (victorlin) [reply augur/filter/subsample.py]: Fixed with b3d91b00a598e8d32bc47d73747447237d8894e8

File: augur/filter/subsample.py [hunk 3]
```diff
@@ -297,6 +299,69 @@ def get_probabilistic_group_sizes(groups, target_group_size, random_seed=None):
...
+                values_for_unweighted_columns[column].add(column_to_value_map[column])
+
+        # 2. Create a DataFrame for all permutations of values in unweighted columns
+        lists = [list(values_for_unweighted_columns[column]) for column in unweighted_columns]
+        unweighted_permutations = pd.DataFrame(list(itertools.product(*lists)), columns=unweighted_columns)
+
+        # 3. Merge with the weights DataFrame so it has all grouping columns
+        weights = pd.merge(unweighted_permutations, weights, how='cross')
+
+    # Drop any groups that don't appear in metadata
```
Comment (jameshadfield) [inline augur/filter/subsample.py:342]: I think this is ok... initially I didn't think this was the right move but using the following example I think it's probably fine. But other people should check this bit!

Example: we want to sample twice as many location A sequences (cf. location B) for each year, but location A is missing data from one year:

```console
$ weights.tsv
year	location	weight
2001	A	2
2001	B	1
2002	A	2
2002	B	1

$ metadata.tsv
strain	date	location
SEQ3	2001	A
SEQ4	2001	B
SEQ5	2002	B
```

Targeting 100 sequences, if we drop groups missing from metadata (this PR, as it stands) we'll target `group_sizes {(2001, 'A'): 50, (2001, 'B'): 25, (2002, 'B'): 25}` and (assuming there's enough entries in the metadata) we'll get that many sequences - note that we now have a 50-50 mix of A vs B samples.

If we don't drop groups, then we'll target `group_sizes {(2001, 'A'): 33, (2001, 'B'): 16, (2002, 'A'): 33, (2002, 'B'): 16}`. The `(2002, 'A')` group will get 0 sequences, so we'll still get a 50-50 mix of A-B samples, just less of them.

**Review by jameshadfield — COMMENTED** [review]
This is going to be really powerful @victorlin, nice work. I tested this out a little bit, but haven't gone through every part of it. There's a few aspects that this brought up for which I'll think through a little more, but I think the overall direction is sound.

File: augur/filter/subsample.py [hunk 4]
```diff
@@ -297,6 +299,69 @@ def get_probabilistic_group_sizes(groups, target_group_size, random_seed=None):
...
+        form) to group sizes
+    """
+    weights = pd.read_csv(weights_file, delimiter='\t')
+
+    weighted_columns = get_weighted_columns(weights_file)
+
+    # Allow other columns in group_by to be equally weighted (uniform sampling)
+    unweighted_columns = list(set(group_by) - set(weighted_columns))
+
+    if unweighted_columns:
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py:321]: Great example! I've pushed it as a test in 740e74883e0d361d08cec97eaffd1560d365106d with an edit to create a metadata file with 250 sequences so `--subsample-max-sequences 100` makes sense.

I think the behavior is expected, and related to what the observation in your [other thread](https://github.com/nextstrain/augur/pull/1454#discussion_r1596227089). I tried to explain it here, but maybe we can walk through some more scenarios on a call.

https://github.com/nextstrain/augur/blob/740e74883e0d361d08cec97eaffd1560d365106d/tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t#L60-L62

File: augur/filter/subsample.py [hunk 5]
```diff
@@ -368,6 +368,12 @@ def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size,
     rng = np.random.default_rng(random_seed)
     weights[SIZE_COLUMN] = (weights[FRACTIONAL_SIZE_COLUMN].add(rng.random(len(weights)))).astype(int)
 
+    # TODO: consider adding an option to output the computed sizes since that
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py]: Added in 2ee8b26fa539b6a9fff17a8e53e6a0866ccd2e73 since it's useful info for the [other thread](https://github.com/nextstrain/augur/pull/1454#discussion_r1596214631).

**Comment by genehack** [comment]
Having just read this from scratch, and hoping to help it get moving again, it seems like the remaining outstanding discussion items are: 

#1 [adding an (optional?) output with requested and calculated group sizes](https://github.com/nextstrain/augur/pull/1454#discussion_r1596196976) – which seems like a pretty valuable debugging tool to me, but maybe non-blocking for this particular PR
#2 [validating that the current behavior when a group-by group is empty](https://github.com/nextstrain/augur/pull/1454#discussion_r1596214631) is desired [convo continued in second comment](https://github.com/nextstrain/augur/pull/1454#discussion_r1596227089) – *seems* like this is the case but maybe a bit more explicit validation/commentary would be good?

Did I miss anything @victorlin ?

File: augur/filter/subsample.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py]: Leaving this open to add **number of outputted sequences**. I don't think it's gonna be trivial to add (properly), but need to dig a little deeper.

**Comment by victorlin** [comment]
Thanks for bumping this @genehack. I added a comment on the thread for (1). Testing and discussion in https://github.com/nextstrain/ncov/pull/1106 should address (2) and impact the overall direction of this PR. I want to make sure we have at least one solid use case for this feature.

File: augur/filter/subsample.py (hunk 3 above)
Comment (victorlin) [reply augur/filter/subsample.py:345]: Assuming weight files are constructed by data sources spanning the full range of possible group-by values (e.g. global population sizes, case counts from all time), dropping groups will be very common and we should make sure we get this right. See https://github.com/nextstrain/ncov/pull/1106 as a real-world example.

File: augur/filter/subsample.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameshadfield) [reply augur/filter/subsample.py:321]: > I think the behavior is expected

I don't agree. Let me use a more extreme (but not unheard of) example. The temporal distribution of samples across geographies can vary greatly - sometimes they're only available for one single year, sometimes every year for decades. Using the same command and weights as before with the following metadata (but imagining for each row there's many entries with the same `{date,location}` metadata so we have plenty of samples to choose from):

```console
strain	date	location
SEQ_A.1	2000	A
SEQ_A.2	2000	A
SEQ_A.3	2000	A
SEQ_B.1	2000	B
SEQ_B.2	2001	B
SEQ_B.3	2002	B
SEQ_B.4	2003	B
SEQ_B.5	2004	B
SEQ_B.6	2005	B
SEQ_B.7	2006	B
```

If I request `--group-by location year` I want around 67 x location A and 33 x location B because of the 2:1 weighting I specified. What I'll actually get is ~22 from location A and 77 from location B. The more unequal the temporal distribution of data is across locations the more the output will deviate from my specified weightings.
 
Put another way, I think extra group-bys should be used to partition the data _within_ the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py:321]: > Put another way, I think extra group-bys should be used to partition the data _within_ the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

This makes sense. I'll look into updating the logic to reflect this.

**Review by genehack — DISMISSED** [review]

File: augur/filter/subsample.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py:322]: Done in 1a9dd7cadbbe899464e94b1aee2cfe7e0bd6d66f. Gonna test on https://github.com/nextstrain/ncov/pull/1106 tomorrow.

File: augur/filter/subsample.py [hunk 6]
```diff
@@ -359,8 +359,14 @@ def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size):
         raise AugurError(error)
 
     # Calculate maximum group sizes based on weights
+    FRACTIONAL_SIZE_COLUMN = '_augur_filter_target_size_fraction'
     SIZE_COLUMN = '_augur_filter_target_size'
-    weights[SIZE_COLUMN] = weights[WEIGHTS_COLUMN] / weights[WEIGHTS_COLUMN].sum() * target_total_size
+    weights[FRACTIONAL_SIZE_COLUMN] = weights[WEIGHTS_COLUMN] / weights[WEIGHTS_COLUMN].sum() * target_total_size
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py]: Renamed in 2d84b0d838911f29a32665c9468e2695aeb9b5d5

File: augur/filter/subsample.py [hunk 7]
```diff
@@ -352,10 +352,20 @@ def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size,
 
     missing_groups = set(groups) - set(weights[group_by].apply(tuple, axis=1))
     if missing_groups:
-        error = "The following groups appear in the metadata but are missing from the weights file:"
-        for group in sorted(missing_groups):
+        n_missing = len(missing_groups)
+        error = f"{n_missing} groups appear in the metadata but are missing from the weights file"
+        missing_weights = pd.DataFrame(sorted(missing_groups), columns=group_by)
+        missing_weights['weight'] = ''
+        missing_weights_file = 'missing_weights.tsv'
```
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py]: Added in 19c685f7c4b265dbabe5b71ff50c67a55b18db4e

**Review by joverlee521 — APPROVED** [review]
I haven't fully grokked what's going on in `get_weighted_group_sizes`, but the behaviors described in the functional tests make sense to me. 

Once this feature is released, it definitely deserves it's own section in the [Filtering and Subsampling docs](https://docs.nextstrain.org/en/latest/guides/bioinformatics/filtering-and-subsampling.html#subsampling-within-augur-filter).

File: augur/filter/weights_file.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+WEIGHTS_COLUMN = 'weight'
+
+
+class InvalidWeightsFile(AugurError):
+    def __init__(self, file, error_message):
+        super().__init__(f"Bad weights file {file!r}.\n{error_message}")
+
+
+def read_weights_file(weights_file):
+    weights = pd.read_csv(weights_file, delimiter='\t')
```
Comment (genehack) [inline augur/filter/weights_file.py:16]: per [Tom's recent PR](https://github.com/nextstrain/augur/pull/1561) maybe set `newline` here?

```suggestion
    weights = pd.read_csv(weights_file, newline='', delimiter='\t')
```

Comment (huddlej) [reply augur/filter/weights_file.py:16]: @genehack I thought Tom's recommendation was focused on the built-in `csv` module and not the pandas `read_csv` function which doesn't have a `newline` argument? [pandas seems to open files internally](https://github.com/pandas-dev/pandas/blob/7ee10913d7192a285383ceec9539191b784c1ed4/pandas/io/formats/format.py#L1048) with `newline=""`.

Comment (genehack) [reply augur/filter/weights_file.py:16]: > [pandas seems to open files internally](https://github.com/pandas-dev/pandas/blob/7ee10913d7192a285383ceec9539191b784c1ed4/pandas/io/formats/format.py#L1048) with `newline=""`.

🤦 you are correct, i over-indexed on `csv`...

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  >   --group-by-weights weights.tsv \
+  >   --subsample-max-sequences 100 \
+  >   --subsample-seed 0 \
+  >   --output-group-by-sizes target_group_sizes.tsv \
+  >   --output-strains strains.txt 2>/dev/null
+
+  $ cat target_group_sizes.tsv
+  location	weight	_augur_filter_target_size_int
+  A	2	67
+  B	1	33
```
Comment (jameshadfield) [inline tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:36]: Any reason not to also check the actual number of strains? e.g. `grep 'A_' < strains.txt | wc -l`. It should be stable for a given `--subsample-seed` I think?

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  2000	B	0.3333333333333333	16
+  2001	A	0.5	25
+  2001	B	0.3333333333333333	16
+  2002	B	0.3333333333333333	17
+
+If a single sequence is added for group (2002,A), the weighting now appears
+"equal" among all years and locations.
+
+However, there is only 1 sequence available in (2002,A), much lower than the
+requested 17, so the total number of sequences outputted is lower than requested.
```
Comment (jameshadfield) [inline tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:86]: This is an interesting example! It's similar to [our previous conversation](https://github.com/nextstrain/augur/pull/1454#discussion_r1596214631) about how to divide up samples within `location=A` via the year axis:

> Put another way, I think extra group-bys should be used to partition the data within the existing weighted groups, and thus the temporal distribution of data from a location will only affect the sampling of data from that location not other locations.

In this context we aim for uniform sampling within "A", so around 16 x 2000, 16 x 2001, 16 x 2002, and since there's only one 2002/A sample we undersample A and thus don't get our desired 1:1 A:B ratio. That's kinda weird. The alternate approach (at least, that I can think of) is to in-fill samples until we get to the target number for A. So we'd start with `{2000: 16, 2001: 16, 2002: 1}` which totals 33, but we want 50, so we'd add iteratively add samples with a ratio of (1/3 x 2000, 1/3 x 2001, 1/3 x 2002) until we reached 50 samples which would be circa `{2000: 25, 2001: 24, 2002: 1}`. 

No matter what approach we use, we should add output describing what's happening. I.e.

```diff
NOTE: Weights were not provided for the column 'year'.
Using equal weights across values in that column
+ NOTE: Within group "location=A" we uniformly sampled across "year"
+ however there were insufficient samples for "year=2002" resulting
+ in an undersampling of "location=A". 
```

```diff
NOTE: Weights were not provided for the column 'year'.
Using equal weights across values in that column
+ NOTE: Within group "location=A" we uniformly sampled across "year"
+ however there were insufficient samples for "year=2002" resulting in
+ in the output distribution not matching the input distribution
```

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t [hunk 11]
```diff
@@ -0,0 +1,109 @@
+Setup
+
+  $ source "$TESTDIR"/_setup.sh
+
```
Comment (jameshadfield) [inline tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:4]: If we add a test which produces 50 x A and 150 x B, like so:

```
  $ echo "strain	date	location" > metadata.tsv
  $ for i in $(seq 1 50); do
  >     echo "2000A_$i	2000	A" >> metadata.tsv
  >     echo "2000B_$i	2000	B" >> metadata.tsv
  >     echo "2001B_$i	2001	B" >> metadata.tsv
  >     echo "2002B_$i	2002	B" >> metadata.tsv
  > done
```

and then ask for a 1:1 ratio of A:B with `--subsample-max-sequences 200` then we target 100xA, 100xB which isn't possible. Not a whole lot we can do here I don't think, but we should at least add a note indicating that certain groups (e.g. "A") didn't have enough samples and thus the output distribution doesn't match the intended weights. 

(This behaviour holds across both `--group-by location` and `--group-by location year`, although the latter can cause extra complications -- see the other review thread.)

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 9 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:36]: How about checking the grouped column values in output metadata instead of strains? Added that in 4f7738daa95c9e68087ffe2bbe81ee5556dbc5c3

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 10 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: This odd behavior already exists for current "uniform" sampling (the behavior when `--group-by-weights` is removed from this example). The weighting is slightly different in that it will weight evenly across all permutations of `year` and `location` values that have at least one sequence in the metadata. In other words, it would be the equivalent of specifying this weights file:

```sh
cat >weights.tsv <<~~
year	location	weight
2000	A	1
2001	A	1
2002	A	1
2000	B	1
2001	B	1
2002	B	1
~~
```

which would, before adding the `2002A_1	2002	A` row, produce this `--output-group-by-sizes` file:

```
year	location	weight	_augur_filter_target_size_int
2000	A			1		20
2001	A			1		20
2000	B			1		20
2001	B			1		20
2002	B			1		20
```

and after adding that row produce this:

```
year	location	weight	_augur_filter_target_size_int
2000	A			1		17
2001	A			1		16
2002	A			1		16
2000	B			1		16
2001	B			1		17
2002	B			1		17
```

This gets back to the point in https://github.com/nextstrain/augur/pull/1454#discussion_r1607249501 about adding number of outputted sequences to this output file. Example:

```
year	location	weight	_augur_filter_target_size_int	sequences_that_passed_filter
2000	A			1		0								20
2000	B			1		0								20
2001	A			1		0								20
2001	B			1		0								20
2002	B			1		0								20
```

```
year	location	weight	_augur_filter_target_size_int	sequences_that_passed_filter
2000	A			1		17								17
2001	A			1		16								16
2002	A			1		16								1
2000	B			1		16								16
2001	B			1		17								17
2002	B			1		17								17
```

The missing piece there (and also what's needed to produce the additional output in your suggestion) is to update this function to be aware of the number of rows per group available in the metadata. It should be doable, I will take a look. Once that information is availalbe we can also consider the in-filling approach you've described.

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 11 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:4]: The solution here is the same as in the other thread, see my comment:

> The missing piece there (and also what's needed to produce the additional output in your suggestion) is to update this function to be aware of the number of rows per group available in the metadata. It should be doable, I will take a look.

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 9 above)
Comment (jameshadfield) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:36]: Perfect!

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 10 above)
Comment (jameshadfield) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: I'm not entirely convinced that in-filling is the best approach, it was just what I came up with upon review, so perhaps don't invest time in that implementation until others have expressed their thoughts. But having warning/log messages printed in these cases will be helpful (assuming one reads the warning messages).

Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: I made a similar note in the description of https://github.com/nextstrain/ncov/pull/1106:

> Caveat: weighted target sizes are calculated without taking into account the actual number of sequences available per group. This means under-sampled countries will still be under-sampled even if it has a high weight, and results in significantly fewer total sequences than requested by `--subsample-max-sequences`. This is already the case with the uniform sampling approach prior to this PR, but it may be more noticeable now with large countries that are under-sampled.

**Comment by victorlin** [comment]
> @joverlee521: I haven't fully grokked what's going on in `get_weighted_group_sizes`

I've split this function up into smaller parts and added types to make it more readable: c305fe8258ecf72747f60e38bdd252a79f38e245

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 10 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: Added better outputs in 3c64691f5e853b844d69c4cfbf7fb765bdb58449. Will test in https://github.com/nextstrain/ncov/pull/1106

Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: This is what it looks like, from [run](https://github.com/nextstrain/ncov/actions/runs/10294116138/job/28491628027#step:5:793):

```
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2017', "month='2017-8'"] but only 1 is available.
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2018', "month='2018-8'"] but only 1 is available.
WARNING: Targeted 18 sequences for group ["country='China'", 'year=2019', "month='2019-10'"] but only 1 is available.
WARNING: Targeted 20 sequences for group ["country='India'", 'year=2020', "month='2020-1'"] but only 2 are available.
WARNING: Targeted 2 sequences for group ["country='Bangladesh'", 'year=2020', "month='2020-2'"] but only 1 is available.
```

It was easy to implement and is more verbose than your initial suggestion, but lacks info on weighted vs. unweighted columns. That could be added:

```
WARNING: Undersampling occurred for the following groups within ('country'='China'):
         - Targeted 19 sequences for ('year'='2017', 'month='2017-8') but only 1 is available
         - Targeted 19 sequences for ('year'='2018', 'month='2018-8') but only 1 is available
         - Targeted 18 sequences for ('year'='2019', 'month='2019-10') but only 1 is available
WARNING: Undersampling occurred for the following groups within ('country'='India'):
         - Targeted 18 sequences for ('year'='2020', 'month='2020-1') but only 2 are available
WARNING: Undersampling occurred for the following groups within ('country'='Bangladesh'):
         - Targeted 2 sequences for ('year'='2020', 'month='2020-2') but only 1 is available
```

The output would be simpler for the [other case](https://github.com/nextstrain/augur/pull/1454#discussion_r1704761006) (weighted columns only):

```
WARNING: Undersampling occurred for ('location'='A'): targeted 100 sequences but only 50 are available.
```

Comment (trvrb) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: I agree that this is a tricky situation. In the original formulation of only `--sequences-per-group` the idea was to say specify `--sequences-per-group 10` and `--group-by country` would target 10 sequences per country and randomly sample these sequences for each country group. In the original formulation, we wouldn't top-up other countries. I think this is a semantic complication with adding the convenience parameter of `--subsample-max-sequences`. I'd think of `--subsample-max-sequences` and solely specifying `--sequences-per-group`.

All this to say, that I believe the current behavior of 
```
WARNING: Targeted 19 sequences for group ["country='China'", 'year=2017', "month='2017-8'"] but only 1 is available.
```
is exactly what we want. It's okay if there's holes in the resulting sampling if the underlying data doesn't support filling all groups. It's a bit hard to predict what feels most natural to everyone, but this behavior seems like the least surprising behavior to me.

File: augur/filter/__init__.py [hunk 12]
```diff
@@ -68,8 +68,32 @@ def register_arguments(parser):
...
+
+        (1) The first row must be a header.
+        (2) There must be a numeric ``weight`` column (weights can take on any
+            non-negative values).
+        (3) Other columns must be a subset of columns used in ``--group-by``,
+            with combinations of values covering all combinations present in the
+            metadata.
+        (4) This option only applies when ``--group-by`` and
+            ``--subsample-max-sequences`` are provided.
+        (5) This option cannot be used with ``--probabilistic-sampling``.
```
Comment (trvrb) [inline augur/filter/__init__.py:84]: I believe that `--probabilistic-sampling` should be compatible with `--group-by-weights`. 

Here is a quick example. Let's say we have countries `A`, `B`, `C`, etc...,  through `J`. We specify `--subsample-max-sequences of 5`. With 10 groups and with 5 total sequences targeted you should get `--sequences-per-group` of 0.5. Without weighted subsampling this would target `A` with 0.5 total sequences, `B` with 0.5 total sequences, etc... 

Now let's say we specify `--group-by-weights weights.tsv` with
```
country weight
A       10
B       2
C       1
D       1
E       1
F       1
G       1
H       1
I       1
J       1
```

With above weights you grab `A` 10/20 = 50% of the time when you grab a sample, you grab `B` 2/20 = 10% of the time when you grab a sample, you grab `C` 1/20 = 5% of the time when you grab a sample, etc...

So, if you specify `--subsample-max-sequences 5` the final targets should be, `A` at 50% × 5 = 2.5 total sequences probabilistically sampled, `B` at 10% × 5 = 0.5 total sequences probabilistically sampled, `C` at 5% × 5 = 0.25 total sequences probabilistically sampled, etc...

This same logic should extend to more granular groups created by `--group-by country year`, etc...

I think it's important that probabilistic sampling eventually work with `--group-by-weights`. However, I'm very okay to merge this PR without support for `--probabilistic-sampling` and follow up soon afterwards with a separate PR that implements this.

**Review by trvrb — APPROVED** [review]
@victorlin: This is great! I appreciate all the documentation and tests. I'm good for this to be merged as it stands. ~However, I do think we should follow up shortly with a separate PR that implements support for `--probabilistic-sampling` + `--group-by-weights` per https://github.com/nextstrain/augur/pull/1454#discussion_r1723933707.~ _I was confused. Nothing more to be done with probabilistic sampling._

File: augur/filter/__init__.py (hunk 12 above)
Comment (victorlin) [reply augur/filter/__init__.py:84]: Good point to raise. Probabilistic sampling is inherent to the weighted sampling implementation:

https://github.com/nextstrain/augur/blob/66d0df4b9c7f5396a7f7e90a7c7e6b08d96a7d2e/augur/filter/subsample.py#L449-L454

<details>
<summary>
Detailed example
</summary>

```sh
cat >metadata.tsv <<~~
strain	country
A1	A
B1	B
C1	C
D1	D
E1	E
F1	F
G1	G
H1	H
I1	I
J1	J
~~

cat >weights.tsv <<~~
country	weight
A	10
B	2
C	1
D	1
E	1
F	1
G	1
H	1
I	1
J	1
~~

augur filter \
  --metadata metadata.tsv \
  --group-by country \
  --group-by-weights weights.tsv \
  --subsample-max-sequences 5 \
  --subsample-seed 0 \
  --output-group-by-sizes target_group_sizes.tsv \
  --output-metadata filtered.tsv
```

Output:

```
WARNING: Asked to provide at most 5 sequences, but there are 10 groups.
Sampling with weights defined by weights.tsv.
WARNING: Targeted 3 sequences for group ["country='A'"] but only 1 is available.
6 strains were dropped during filtering
        6 were dropped because of subsampling criteria
4 strains passed all filters
```
`target_group_sizes.tsv` looks like:

```
cat target_group_sizes.tsv | tsv-pretty 
country  weight  _augur_filter_target_size  _augur_filter_input_size  _augur_filter_output_size
A            10                          3                         1                          1
B             2                          0                         1                          0
C             1                          0                         1                          0
D             1                          0                         1                          0
E             1                          1                         1                          1
F             1                          1                         1                          1
G             1                          0                         1                          0
H             1                          0                         1                          0
I             1                          0                         1                          0
J             1                          1                         1                          1
```

</details>


However, it uses a different approach than [get_probabilistic_group_sizes](https://github.com/nextstrain/augur/blob/d8faf01ec1008b4e7b98fa2285d2f9c931c505d7/augur/filter/subsample.py#L252) which instead samples from a Poisson distribution. I think it might be possible to replace that with the approach I used here, but I'll investigate that separately.

TLDR: I've worded this poorly and probabilistic sampling is always used. I'll update it to say "this option cannot be used with ``--no-probabilistic-sampling``".

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 11 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:4]: Addressed with warning in output. Example:

```
WARNING: Targeted 3 sequences for group ["country='A'"] but only 1 is available.
```

File: tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t (hunk 10 above)
Comment (victorlin) [reply tests/functional/filter/cram/subsample-weighted-and-uniform-mix.t:93]: Great, I'll leave the output as-is. We can tweak as we start using this in workflows.

File: augur/filter/subsample.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (victorlin) [reply augur/filter/subsample.py]: This was implemented for https://github.com/nextstrain/augur/pull/1454#discussion_r1708166896

File: augur/filter/__init__.py (hunk 12 above)
Comment (trvrb) [reply augur/filter/__init__.py:84]: Ah! Awesome. This makes sense. I just tested this exact scenario and everything works as I expected with "probabilistic sampling". If you could update the help text as you say, this would be a bit less confusing.

Comment (victorlin) [reply augur/filter/__init__.py]: Help text updated:

https://github.com/nextstrain/augur/blob/8e3d6be81559a25c41ce1d98342a5ac458bffdd7/augur/filter/__init__.py#L85

Good that you caught this. It required some additional tweaks and proper validation:

https://github.com/nextstrain/augur/blob/8e3d6be81559a25c41ce1d98342a5ac458bffdd7/augur/filter/validate_arguments.py#L61-L65

**Comment by victorlin** [comment]
Finished addressing all conversations and cleaned up commits. I'm going to re-run trial builds once more on https://github.com/nextstrain/ncov/pull/1106 with latest changes then merge this tomorrow if all looks good.
